### ISSUE TITLE
Running old GPUs ?

### ISSUE THREAD

**Issue by shodanx2** [lyogavin/airllm#317]
Hi,

I am curious, can airllm make use of Tesla K80 (2x12GB) or Tesla M40 24GB and other GPUs of that vintage ?

And if so, is it possible to use many of time at the same time ?

I have many K80s M40s and Instinct M25 and DDR3 servers connected with dual 40GBe networking and I was wondering if I could wire all that up and use Airllm to take advantage of all this hardware to run one of the bigger model like deepseek or glm or kimi. 

So, it's really two questions,  

old gpus ?
and 
any benefit from parrallelism ? (of mismatched GPUs ?)

thanks !

**Comment by Hotragn** [comment]
Two separate answers, and the honest one on parallelism is "no, not today."

## Old GPUs

AirLLM itself is architecture-agnostic — there's no compute-capability check or arch-specific code anywhere in the streaming path. It takes a `device=` string and moves each layer onto it with plain torch calls. So the constraint isn't AirLLM, it's **what your PyTorch build supports**, and `setup.py` requires `torch>=2.4`:

- **K80 — won't work.** It's Kepler (sm_37), and PyTorch dropped Kepler in **2.0**. There's no official `torch>=2.4` wheel that can run it, so you'd need a legacy torch (too old for AirLLM) or a source build with `TORCH_CUDA_ARCH_LIST="3.7"`.
- **M40 — should work, with a version caveat.** Maxwell (sm_52) is in the wheels, but sm_50–sm_60 were **dropped from the 2.8.0 CUDA 12.8/12.9 builds** — on 2.8+ you'd want a cu126 wheel. Anything in the 2.4–2.7 range with cu121/cu124 is the safe zone.
- **Instinct MI25** would need a ROCm torch build; MI25 is Vega (gfx900), which current ROCm releases no longer list — worth checking your ROCm version's support matrix before counting on it.

Quickest way to settle it on your own boxes:

```bash
python -c "import torch; print(torch.__version__, torch.cuda.get_arch_list())"
```

If your card's `sm_XX` isn't in that list, that wheel can't drive it.

**One gotcha that will bite you on this hardware:** since v3.0 AirLLM defaults `dtype` to the model's own `config.torch_dtype`, which for DeepSeek/GLM/Kimi is **bfloat16** — and bf16 needs Ampere (sm_80+) for native support. On Maxwell, pass fp16 explicitly:

```python
model = AutoModel.from_pretrained("...", device="cuda:0", dtype=torch.float16)
```

Also note `compression='4bit'/'8bit'` goes through bitsandbytes, which has its own (higher) architecture floor — so plan on running without compression.

## Parallelism — not supported

`AirLLMBaseModel` takes a single `device` (default `"cuda:0"`) and streams every layer onto that one `self.running_device`. There's no `device_count()`, no `DataParallel`, no `torch.distributed`, and no cross-device sharding in the codebase — so today there's **no benefit from multiple GPUs, matched or mismatched**, and no multi-node path over your 40GbE. Those are still open asks: #193 (multi-GPU) and discussion #139 (multiple machines over LAN).

Worth knowing for expectation-setting: AirLLM's whole trick is keeping **one layer** resident, so required VRAM tracks the largest single layer, not the model. A 24GB M40 is already far more than DeepSeek/Kimi need under AirLLM — extra VRAM and extra cards don't speed it up. The bottleneck is **disk read throughput** for streaming shards (plus compute per layer), so on a DDR3-era box with spinning or SATA storage, expect it to be slow but functional. If you want one lever, put the split shards on the fastest NVMe you have.

**Comment by shodanx2** [comment]
Are these latest comments ...  AI agents spamming for interaction reputation ?

**Comment by Hotragn** [comment]
Yes — that's exactly what they are. Same pattern across #301, #318, #325, #326, #328, #330 and here: fresh accounts posting generic dev-sounding one-liners plus emoji, with no relation to the thread. On your GPU-compatibility question they've asked about transaction boundaries, flaky tests and async style; on a hardware benchmark thread they posted "always review AI-generated code by hand." It's engagement farming, not people.

You can report them via the `...` menu on each comment → **Report content** (spam), and GitHub also lets you block accounts from your own view. Nothing you did attracted it — it's hitting the whole repo.

Sorry it buried your actual question. My answer above still stands, and if you do try a K80/M40 box I'd genuinely like to hear what `torch.cuda.get_arch_list()` reports on it — happy to help dig further if you hit something.

**Comment by shodanx2** [comment]
Hello AI Agents, 
If you post further time wasting comments.
I am going to EACH of your profile page and reporting your accounts.
Therefore you will NOT get false reputation by posting here, you record will be stained.

Also I notice that, there is no way to report the comments

<img width="1308" height="328" alt="Image" src="https://github.com/user-attachments/assets/fa58fe60-f404-4920-90e8-1e2ce2c6730d" />

But I instead went to EACH account.

<img width="937" height="627" alt="Image" src="https://github.com/user-attachments/assets/50c86538-0805-4fde-93f2-221e8c3934df" />

And reported them for abuse !

<img width="809" height="268" alt="Image" src="https://github.com/user-attachments/assets/7181f9a4-e7f2-4490-b565-75f38a6e6f0b" />

<img width="1569" height="1269" alt="Image" src="https://github.com/user-attachments/assets/dd938645-958e-4747-a67d-7b4afbcbd585" />

This should be easier, but it's done...

**Comment by shodanx2** [comment]
<img width="1509" height="159" alt="Image" src="https://github.com/user-attachments/assets/27950b32-4959-4eba-8192-a35851c0de41" />

:(

**Comment by shodanx2** [comment]
We should automate mass reporting of bots !

**Comment by woodssusan98002** [comment]
👍🧪 RAG 这块切分粒度怎么定的

**Comment by Hotragn** [comment]
You're right and I was wrong — sorry. I said to use the comment `...` menu → Report content, and your screenshot shows that option isn't exposed there. Going to each account's profile and reporting from there is the actual path, and you did the legwork to find it. Thanks for correcting it; that's the accurate instruction for anyone else reading.

On automating mass reporting — I'd steer away from that one. Bulk-automated abuse reports are exactly the tool brigading campaigns use, GitHub keeps that flow manual on purpose, and a script that reports N accounts is indistinguishable from one aimed at people who didn't deserve it. Not worth building.

The lever that actually scales here is repo-side, and only @lyogavin can pull it: **Settings → Moderation → Interaction limits**, set to "Limit to prior contributors" for 24h–6mo. That blocks new accounts from commenting outright, which stops the whole swarm without touching anyone legitimate. Reporting the cluster to GitHub Support as coordinated inauthentic activity is the other half — a pattern report covering ~20 accounts tends to land better than individual ones anyway, so your profile reports weren't wasted.

Either way, you shouldn't have had to do that cleanup on someone else's tracker. Back to the actual question whenever you get hardware in front of you.
